File: backend/app/api/asyncapi.py
"""
event7 - AsyncAPI API Routes
Génération, récupération, édition et import des specs AsyncAPI.

Placement: backend/app/api/asyncapi.py

Changelog:
- v1: Generate, get, update, delete, yaml export
- v2: Import preview + apply endpoints
"""

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse

import yaml

from app.api.dependencies import get_schema_service
from app.models.auth import UserContext
from app.models.asyncapi import (
    AsyncAPISpec,
    AsyncAPIGenerateRequest,
    AsyncAPIImportRequest,
    AsyncAPIImportPreview,
    AsyncAPIImportResult,
)
from app.services.schema_service import SchemaService
from app.services.asyncapi_service import AsyncAPIService
from app.services.asyncapi_import_service import AsyncAPIImportService
from app.utils.auth import get_current_user

router = APIRouter(prefix="/api/v1/registries/{registry_id}", tags=["asyncapi"])


# ── Dependency: AsyncAPIService ──

def _get_asyncapi_service(
    service: SchemaService = Depends(get_schema_service),
) -> AsyncAPIService:
    """Construit un AsyncAPIService depuis le SchemaService injecté."""
    return AsyncAPIService(
        provider=service.provider,
        cache=service.cache,
        db=service.db,
        registry_id=service.registry_id,
        registry_url=getattr(service, "registry_url", ""),
    )


# ── Dependency: AsyncAPIImportService ──

def _get_import_service(
    service: SchemaService = Depends(get_schema_service),
) -> AsyncAPIImportService:
    """Construit un AsyncAPIImportService depuis le SchemaService injecté."""
    return AsyncAPIImportService(
        provider=service.provider,
        cache=service.cache,
        db=service.db,
        registry_id=service.registry_id,
    )


# ================================================================
# GENERATE
# ================================================================


@router.post("/subjects/{subject}/asyncapi/generate", response_model=AsyncAPISpec)
async def generate_asyncapi(
    subject: str,
    params: AsyncAPIGenerateRequest | None = None,
    user: UserContext = Depends(get_current_user),
    asyncapi_service: AsyncAPIService = Depends(_get_asyncapi_service),
):
    """Génère une spec AsyncAPI depuis le schema + enrichments."""
    try:
        return await asyncapi_service.generate(subject, params, user_id=str(user.user_id))
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"AsyncAPI generation failed: {e}")


# ================================================================
# GET / UPDATE / DELETE
# ================================================================


@router.get("/subjects/{subject}/asyncapi", response_model=AsyncAPISpec)
async def get_asyncapi(
    subject: str,
    user: UserContext = Depends(get_current_user),
    asyncapi_service: AsyncAPIService = Depends(_get_asyncapi_service),
):
    """Récupère la spec AsyncAPI existante."""
    spec = await asyncapi_service.get_spec(subject)
    if not spec:
        raise HTTPException(status_code=404, detail=f"No AsyncAPI spec found for {subject}")
    return spec


@router.put("/subjects/{subject}/asyncapi", response_model=AsyncAPISpec)
async def update_asyncapi(
    subject: str,
    spec_content: dict,
    user: UserContext = Depends(get_current_user),
    asyncapi_service: AsyncAPIService = Depends(_get_asyncapi_service),
):
    """Met à jour manuellement une spec AsyncAPI."""
    try:
        return await asyncapi_service.update_spec(
            subject, spec_content, user_id=str(user.user_id)
        )
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Update failed: {e}")


@router.delete("/subjects/{subject}/asyncapi")
async def delete_asyncapi(
    subject: str,
    user: UserContext = Depends(get_current_user),
    asyncapi_service: AsyncAPIService = Depends(_get_asyncapi_service),
):
    """Supprime une spec AsyncAPI."""
    deleted = asyncapi_service.db.delete_asyncapi_spec(
        asyncapi_service.registry_id, subject, user_id=str(user.user_id)
    )
    if not deleted:
        raise HTTPException(status_code=404, detail=f"No AsyncAPI spec found for {subject}")
    return {"ok": True}


# ================================================================
# YAML EXPORT
# ================================================================


@router.get("/subjects/{subject}/asyncapi/yaml")
async def export_asyncapi_yaml(
    subject: str,
    user: UserContext = Depends(get_current_user),
    asyncapi_service: AsyncAPIService = Depends(_get_asyncapi_service),
):
    """Exporte la spec en YAML."""
    spec = await asyncapi_service.get_spec(subject)
    if not spec:
        raise HTTPException(status_code=404, detail=f"No AsyncAPI spec found for {subject}")

    yaml_content = yaml.dump(spec.spec_content, default_flow_style=False, allow_unicode=True)
    return {"content": yaml_content}


# ================================================================
# IMPORT — Preview (dry-run) & Apply (persist)
# ================================================================


@router.post("/asyncapi/import/preview", response_model=AsyncAPIImportPreview)
async def import_preview(
    payload: AsyncAPIImportRequest,
    user: UserContext = Depends(get_current_user),
    import_service: AsyncAPIImportService = Depends(_get_import_service),
):
    """Parse an AsyncAPI spec and preview what would be created.
    
    Nothing is persisted — this is a dry-run.
    Returns channels, bindings, enrichments, unknown schemas, and warnings.
    """
    try:
        return await import_service.preview(payload)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Import preview failed: {e}")


@router.post("/asyncapi/import/apply", response_model=AsyncAPIImportResult)
async def import_apply(
    payload: AsyncAPIImportRequest,
    user: UserContext = Depends(get_current_user),
    import_service: AsyncAPIImportService = Depends(_get_import_service),
):
    """Parse an AsyncAPI spec and persist all extracted entities.
    
    Creates channels, bindings, enrichments, and optionally registers
    unknown schemas in the Schema Registry.

    Set register_schemas=true to push schemas not found in the SR.
    """
    try:
        return await import_service.apply(payload)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Import failed: {e}")

File: frontend/src/lib/api/asyncapi.ts
// src/lib/api/asyncapi.ts
// AsyncAPI API module — wraps event7 backend routes
// v2: Added import preview + apply functions
//
// Placement: frontend/src/lib/api/asyncapi.ts

import { api } from "./client";
import type {
  AsyncAPISpec,
  AsyncAPIGenerateRequest,
  AsyncAPIYamlExport,
  AsyncAPIImportRequest,
  AsyncAPIImportPreview,
  AsyncAPIImportResult,
  AsyncAPIOverviewResponse
} from "@/types/asyncapi";

// ════════════════════════════════════════════════════════════════════
// GENERATE / VIEW / EDIT (existing)
// ════════════════════════════════════════════════════════════════════

/** Generate an AsyncAPI spec from a subject's schema + enrichments. */
export async function generateAsyncAPI(
  registryId: string,
  subject: string,
  params?: AsyncAPIGenerateRequest
): Promise<AsyncAPISpec> {
  return api.post<AsyncAPISpec>(
    `/api/v1/registries/${registryId}/subjects/${encodeURIComponent(subject)}/asyncapi/generate`,
    params ?? {}
  );
}

/** Get an existing AsyncAPI spec (from cache/DB). */
export async function getAsyncAPI(
  registryId: string,
  subject: string
): Promise<AsyncAPISpec | null> {
  try {
    return await api.get<AsyncAPISpec>(
      `/api/v1/registries/${registryId}/subjects/${encodeURIComponent(subject)}/asyncapi`
    );
  } catch {
    return null;
  }
}

/** Update an AsyncAPI spec manually. */
export async function updateAsyncAPI(
  registryId: string,
  subject: string,
  specContent: Record<string, unknown>
): Promise<AsyncAPISpec> {
  return api.put<AsyncAPISpec>(
    `/api/v1/registries/${registryId}/subjects/${encodeURIComponent(subject)}/asyncapi`,
    specContent
  );
}

/** Export the spec as YAML string. */
export async function exportAsyncAPIYaml(
  registryId: string,
  subject: string
): Promise<string> {
  const data = await api.get<AsyncAPIYamlExport>(
    `/api/v1/registries/${registryId}/subjects/${encodeURIComponent(subject)}/asyncapi/yaml`
  );
  return data.content;
}

// ════════════════════════════════════════════════════════════════════
// IMPORT (new)
// ════════════════════════════════════════════════════════════════════

/**
 * Preview what would be created from an AsyncAPI spec (dry-run).
 * Nothing is persisted.
 */
export async function importPreview(
  registryId: string,
  specContent: Record<string, unknown>
): Promise<AsyncAPIImportPreview> {
  return api.post<AsyncAPIImportPreview>(
    `/api/v1/registries/${registryId}/asyncapi/import/preview`,
    { spec_content: specContent } satisfies AsyncAPIImportRequest
  );
}

/**
 * Parse an AsyncAPI spec and persist all extracted entities.
 * Creates channels, bindings, enrichments, and optionally registers schemas.
 */
export async function importApply(
  registryId: string,
  specContent: Record<string, unknown>,
  registerSchemas: boolean = false
): Promise<AsyncAPIImportResult> {
  return api.post<AsyncAPIImportResult>(
    `/api/v1/registries/${registryId}/asyncapi/import/apply`,
    { spec_content: specContent, register_schemas: registerSchemas } satisfies AsyncAPIImportRequest
  );
}

// ════════════════════════════════════════════════════════════════════
// OVERVIEW (Dual Mode)
// ════════════════════════════════════════════════════════════════════

/** Fetch dual-mode overview: KPIs + per-subject status. */
export async function getAsyncAPIOverview(
  registryId: string
): Promise<AsyncAPIOverviewResponse> {
  return api.get<AsyncAPIOverviewResponse>(
    `/api/v1/registries/${registryId}/asyncapi/overview`
  );
}
 